piet-gpu-hal/examples/shader/stack.comp

```diff
@@ -0,0 +1,217 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Implementation of stack monoid (companion to blog post)
+
+#version 450
+
+#define LG_PART_SIZE 4
+#define PART_SIZE (1 << LG_PART_SIZE)
+
+layout(local_size_x = PART_SIZE) in;
+
+layout(set = 0, binding = 0) buffer DataBuf {
+    uint[] data;
+};
+
+layout(set = 0, binding = 1) volatile buffer StateBuf {
+    uint[] state;
+};
+
+#define STATE_COUNTER 0
```

Comment:
Is STATE_COUNTER global? If so, I suggest replacing it with a `state_counter` field in StateBuf.